SimMuon/GEMDigitizer/src/ME0ReDigiProducer.cc

```diff
@@ -19,259 +19,385 @@
+
+			//Digis store sigmaX,sigmaY, correlationCoef
+			const float sigmaX = std::sqrt(digiLocalError.xx());
+			const float sigmaY = std::sqrt(digiLocalError.yy());
+			const float corrCoef = digiLocalError.xy() /(sigmaX*sigmaY);
+
+			//Fill in the new collection
+			ME0DigiPreReco out_digi(digiLocalPoint.x(), digiLocalPoint.y(),
+					sigmaX, sigmaY, corrCoef, tof, digi->pdgid(), digi->prompt());
+			output_digis.insertDigi(me0Id, out_digi);
+
+			// store index of previous detid and digi
+			output_digimap.insertDigi(me0Id, newDigiIdx);
+			newDigiIdx++;
+
+			LogTrace("ME0ReDigiProducer::buildDigis") << "("<<digiLocalPoint.x()<<","<<digiLocalPoint.y()<<","<<sigmaX<<","<<sigmaY<<","<< tof<<") ";
+		}
+
+		chDigiMap.clear();
+
+
 	}
-	else {	 
-	  // roll changed, get new ME0EtaPartition
-	  newPart = geometry_->etaPartition(ME0DetId(detId.region(), detId.layer(), detId.chamber(), newRoll));
+
+}
+
+void ME0ReDigiProducer::fillCentralTOFs() {
+	const auto* mainChamber = geometry->chambers().front();
+	const unsigned int nLayers = mainChamber->nLayers();
+	//Get TOF at center of each partition
+	tofs.clear();
+	tofs.resize(nLayers);
+	LogDebug("ME0ReDigiProducer::fillCentralTOFs()") << "TOF numbers [layer][partition]: " ;
+	for(unsigned int iL = 0; iL < nLayers; ++iL){
+		const auto* layer = mainChamber->layers()[iL];
+		const unsigned int mapLayIDX = layer->id().layer() -1;
+		const unsigned int nPartitions = layer->nEtaPartitions();
+		if(!nPartitions)
+			throw cms::Exception("Setup") << "ME0ReDigiProducer::fillCentralTOFs() - ME0Layer has no partitions.";
+		tofs[mapLayIDX].resize(nPartitions);
+		for(unsigned int iP = 0; iP < nPartitions; ++iP){
+			const unsigned int mapPartIDX = layer->etaPartitions()[iP]->id().roll() -1;
+			const GlobalPoint centralGP(layer->etaPartitions()[iP]->position());
+			tofs[mapLayIDX][mapPartIDX] = (centralGP.mag() / 29.9792); //speed of light [cm/ns]
+			LogDebug("ME0ReDigiProducer::fillCentralTOFs()") << "["<<mapLayIDX<<"]["<<mapPartIDX<<"]="<< tofs[mapLayIDX][mapPartIDX] <<" "<<std::endl;
+		}
 	}
+}
+int ME0ReDigiProducer::getCustomStripProperties(const ME0DetId& detId,const ME0DigiPreReco* inDigi, float& tof,int& strip,  LocalPoint&  digiLocalPoint, LocalError&  digiLocalError ) const {
+	const unsigned int partIdx = tempGeo->findEtaPartition(inDigi->y());
+	LogTrace("ME0ReDigiProducer::buildDigis") << partIdx <<" ";
+	const float partMeanTof = tempGeo->getCentralTOF(detId,partIdx);
+
+	//convert to relative to partition
+	tof -= partMeanTof;
+
+	//get coordinates and errors
+	const float partCenter = tempGeo->getPartCenter(partIdx);
+	const auto* topo = tempGeo->getTopo(partIdx);
+
+	//find channel
+	const LocalPoint partLocalPoint(inDigi->x(), inDigi->y() - partCenter ,0.);
+	strip = topo->channel(partLocalPoint);
+	const float stripF = float(strip)+0.5;
+
+	//get digitized location
+	LocalPoint  digiPartLocalPoint = topo->localPosition(stripF);
+	digiLocalError = topo->localError(stripF, 1./sqrt(12.));
```

Comment:
The std dev. of a flat distribution with length L is L/sqrt(12). The strip topology expects the error in units of strips. I added a comment.